File: tictactoeex/lib/tictactoe.ex
defmodule TicTacToe do
  def winner(board) do 
    row1 = Enum.at(board, 0)
    row2 = Enum.at(board, 1)
    row3 = Enum.at(board, 2)

    cond do
      # Row Checks
      Enum.at(row1, 0) == "o" && Enum.at(row1, 1) == "o" && Enum.at(row1, 2) == "o" -> "o"
      Enum.at(row2, 0) == "o" && Enum.at(row2, 1) == "o" && Enum.at(row2, 2) == "o" -> "o"
      Enum.at(row3, 0) == "o" && Enum.at(row3, 1) == "o" && Enum.at(row3, 2) == "o" -> "o"
      Enum.at(row1, 0) == "x" && Enum.at(row1, 1) == "x" && Enum.at(row1, 2) == "x" -> "x"
      Enum.at(row2, 0) == "x" && Enum.at(row2, 1) == "x" && Enum.at(row2, 2) == "x" -> "x"
      Enum.at(row3, 0) == "x" && Enum.at(row3, 1) == "x" && Enum.at(row3, 2) == "x" -> "x"

      # Column Checks

      Enum.at(row1, 0) == "o" && Enum.at(row2, 0) == "o" && Enum.at(row3, 0) == "o" -> "o"
      Enum.at(row1, 1) == "o" && Enum.at(row2, 1) == "o" && Enum.at(row3, 1) == "o" -> "o"
      Enum.at(row1, 2) == "o" && Enum.at(row2, 2) == "o" && Enum.at(row3, 2) == "o" -> "o"
      Enum.at(row1, 0) == "x" && Enum.at(row2, 0) == "x" && Enum.at(row3, 0) == "x" -> "x"
      Enum.at(row1, 1) == "x" && Enum.at(row2, 1) == "x" && Enum.at(row3, 1) == "x" -> "x"
      Enum.at(row1, 2) == "x" && Enum.at(row2, 2) == "x" && Enum.at(row3, 2) == "x" -> "x"

      # Diagonal Checks

      Enum.at(row1, 0) == "o" && Enum.at(row2, 1) == "o" && Enum.at(row3, 2) == "o" -> "o"
      Enum.at(row1, 2) == "o" && Enum.at(row2, 1) == "o" && Enum.at(row3, 0) == "o" -> "o"
      Enum.at(row1, 0) == "x" && Enum.at(row2, 1) == "x" && Enum.at(row3, 2) == "x" -> "x"
      Enum.at(row1, 2) == "x" && Enum.at(row2, 1) == "x" && Enum.at(row3, 0) == "x" -> "x"

      true -> "draw"
    end
  end
end

File: tictactoe/tic_tac_toe.rb
class TicTacToe
  def initialize(board)
    @board = board
  end

  def winner
    row1 = @board[0]
    row2 = @board[1]
    row3 = @board[2]

    # row checks

    if @board.any? { |row| row.all? {|entry| entry == 'o'} }
      return "o"
    end
    
    if @board.any? { |row| row.all? {|entry| entry == 'x'} }
      return "x"
    end

    # column checks

    if @board.all? {|row| row[0] == 'o'}
      return "o"
    end

    if @board.all? {|row| row[1] == 'o'}
      return "o"
    end

    if @board.all? {|row| row[2] == 'o'}
      return "o"
    end

    if @board.all? {|row| row[0] == 'x'}
      return "x"
    end

    if @board.all? {|row| row[1] == 'x'} # row1[1] == "x" && row2[1] == "x" && row3[1] == "x"
      return "x"
    end

    if @board.all? {|row| row[2] == 'x'} #row1[2] == "x" && row2[2] == "x" && row3[2] == "x"
      return "x"
    end

    # diagonal checks

    if @board.each_with_index.all? {|row, column| row[column] == 'o' }
      return "o"
    end

    if @board.each_with_index.all? {|row, column| row[@board.length - column - 1] == 'o' }
      return "o"
    end

    if @board.each_with_index.all? {|row, column| row[column] == 'x' }
      return "x"
    end

    if @board.each_with_index.all? {|row, column| row[@board.length - column - 1] == 'x' }
      return "x"
    end

    return "draw"
  end
end
